src/mcu.cpp

```diff
@@ -1184,22 +1207,35 @@ void MCU_CloseAudio(void)
 {
     SDL_CloseAudio();
     if (sample_buffer) free(sample_buffer);
+    else if (sample_buffer_float) free(sample_buffer_float);
 }
 
 void MCU_PostSample(int *sample)
 {
-    sample[0] >>= 15;
-    if (sample[0] > INT16_MAX)
-        sample[0] = INT16_MAX;
-    else if (sample[0] < INT16_MIN)
-        sample[0] = INT16_MIN;
-    sample[1] >>= 15;
-    if (sample[1] > INT16_MAX)
-        sample[1] = INT16_MAX;
-    else if (sample[1] < INT16_MIN)
-        sample[1] = INT16_MIN;
-    sample_buffer[sample_write_ptr + 0] = sample[0];
-    sample_buffer[sample_write_ptr + 1] = sample[1];
+    if (sample_buffer)
+    {
+        sample[0] >>= 15;
+        if (sample[0] > INT16_MAX)
+            sample[0] = INT16_MAX;
+        else if (sample[0] < INT16_MIN)
+            sample[0] = INT16_MIN;
+        sample[1] >>= 15;
+        if (sample[1] > INT16_MAX)
+            sample[1] = INT16_MAX;
+        else if (sample[1] < INT16_MIN)
+            sample[1] = INT16_MIN;
+        sample_buffer[sample_write_ptr + 0] = sample[0];
+        sample_buffer[sample_write_ptr + 1] = sample[1];
+    }
+    else if (sample_buffer_float)
+    {
+        const float divRec = 1 / 32768.0f;
+        sample[0] >>= 14;
```

Comment:
There should be no need for bit shift if you divide for `2147483648.0` instead, and you could also keep more precision

Reply:
HI,
If you inspect sample[0] and sample[1] at that place you will notice that if you right shift with 13 you will get samples that has a granulate of 2 if you right shift with 12 you will get  agranulate of 4 and so on. So currently with right shift 14 you get the maximum resolution. So you will not get more resolution with fewer right shifts. 
Moreover the float reolution is not infinite. By dividing with 2^31 you will get less precise results than with exact right shifts and then dividing with 2^15 since 2^15 and its multiplicants can be exactly represented in 32-bit float.

Reply:
makes sense, thanks for the explanation!

Reply:
Hi,
Actually you were right in the sense that the 2 shifts can be skipped and the result is the same. So no more precision, but somewhat better performance. It's because you do not have to divide by 2147483648.0 but only 536870912.0.

So overall this produces the same numbers as the current code:
```
void MCU_PostSample(int *sample)
{
    const float divRec = 1 / 536870912.0f;
    sample_buffer[sample_write_ptr + 0] = sample[0] * divRec;
    sample_buffer[sample_write_ptr + 1] = sample[1] * divRec;     

    sample_write_ptr = (sample_write_ptr + 2) % audio_buffer_size;
} 
```